File: isegoria/src/main/java/com/eulersbridge/isegoria/util/data/RetrofitLiveData.java
package com.eulersbridge.isegoria.util.data;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RetrofitLiveData<T> extends LiveData<T> implements Callback<T> {

    private final @NonNull Call<T> call;

    public RetrofitLiveData(@NonNull Call<T> call) {
        this.call = call;
    }

    @Override
    protected void onActive() {
        if (!call.isCanceled() && !call.isExecuted())
            call.enqueue(this);
    }


    @Override
    public void onResponse(Call<T> call, Response<T> response) {
        if (response.isSuccessful()) {
            postValue(response.body());

        } else {
            postValue(null);
        }
    }

    @Override
    public void onFailure(Call<T> call, Throwable t) {
        setValue(null);
    }

    public void cancel() {
        if (!call.isCanceled())
            call.cancel();
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/util/data/RetrofitLiveData.kt
package com.eulersbridge.isegoria.util.data

import android.arch.lifecycle.LiveData

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RetrofitLiveData<T>(private val call: Call<T>) : LiveData<T>(), Callback<T> {

    override fun onActive() {
        if (!call.isCanceled && !call.isExecuted)
            call.enqueue(this)
    }

    override fun onResponse(call: Call<T>, response: Response<T>) {
        if (response.isSuccessful) {
            postValue(response.body())

        } else {
            postValue(null)
        }
    }

    override fun onFailure(call: Call<T>, t: Throwable) {
        value = null
    }

    fun cancel() {
        if (!call.isCanceled)
            call.cancel()
    }
}
